src/compositemark/callout.ts

```diff
@@ -0,0 +1,131 @@
+import {Config} from '../config';
+import {Encoding} from '../encoding';
+import * as log from '../log';
+import {MarkConfig, MarkDef} from '../mark';
+import {GenericUnitSpec, LayerSpec} from '../spec';
+
+
+export const CALLOUT: 'callout' = 'callout';
+export type CALLOUT = typeof CALLOUT;
+
+export interface CalloutDef {
+  type: CALLOUT;
+  calloutAngle?: number;
+  calloutOffset?: number;
+  calloutLength?: number;
+  labelOffset?: number;
+  callout?: MarkDef;
+  label?: MarkDef;
+}
+
+export function isCalloutDef(mark: CALLOUT | CalloutDef): mark is CalloutDef {
+  return !!mark['type'];
+}
+
+export type CalloutStyle = 'callout-label' | 'callout-rule';
+
+export const CALLOUT_STYLES: CalloutStyle[] = ['callout-label', 'callout-rule'];
+
+export interface CalloutConfig extends MarkConfig {
+  calloutAngle?: number;
+  calloutOffset?: number;
+  calloutLength?: number;
+  labelOffset?: number;
+}
+
+export interface CalloutConfigMixins {
+  /**
+   * Callout Rule Config
+   * @hide
+   */
+  callout?: CalloutConfig;
+}
+
+export const VL_ONLY_CALLOUT_CONFIG_PROPERTY_INDEX: {
+  [k in keyof CalloutConfigMixins]?: (keyof CalloutConfigMixins[k])[]
+} = {
+  callout: ['calloutAngle', 'calloutOffset', 'calloutLength', 'labelOffset']
+};
+
+export function normalizeCallout(spec: GenericUnitSpec<Encoding<string>, CALLOUT>, config: Config): LayerSpec {
+  // TODO:  determine what's the general rule for applying selection for composite marks
+  const {mark: mark, selection: _sel, projection: _p, encoding, ...outerSpec} = spec;
+  if (!isCalloutDef(mark)) {
+    return null;
+  }
+
+  const calloutAngle: number = mark.calloutAngle ? mark.calloutAngle : config.callout.calloutAngle;
+  const calloutOffset: number = mark.calloutOffset ? mark.calloutOffset : config.callout.calloutOffset;
+  const calloutLength: number = mark.calloutLength ? mark.calloutLength : config.callout.calloutLength;
+  const labelOffset: number = mark.labelOffset ? mark.labelOffset : config.callout.labelOffset;
+
+  const calloutOffsetCoordinate: {x: number, y: number} = getCoordinateFromAngleAndLength(calloutAngle, calloutOffset);
+  const calloutLengthCoordinate: {x: number, y: number} = getCoordinateFromAngleAndLength(calloutAngle, calloutOffset + calloutLength);
+  const labelOffsetCoordinate: {x: number, y: number} = getCoordinateFromAngleAndLength(calloutAngle, calloutOffset + calloutLength + labelOffset);
```

Comment:
this is slightly from the input labelOffset though,  maybe `labelTotalOffsetCoor`?